src/test/java/org/opensearch/knn/indices/ModelIndexTests.java

```diff
@@ -0,0 +1,226 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.knn.indices;
+
+import org.opensearch.ExceptionsHelper;
+import org.opensearch.ResourceAlreadyExistsException;
+import org.opensearch.action.ActionListener;
+import org.opensearch.action.admin.indices.create.CreateIndexResponse;
+import org.opensearch.action.delete.DeleteResponse;
+import org.opensearch.action.index.IndexResponse;
+import org.opensearch.index.engine.VersionConflictEngineException;
+import org.opensearch.knn.KNNSingleNodeTestCase;
+import org.opensearch.knn.index.util.KNNEngine;
+import org.opensearch.rest.RestStatus;
+
+import java.io.IOException;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+public class ModelIndexTests extends KNNSingleNodeTestCase {
+
+    public void testCreateModelIndex() throws IOException, InterruptedException {
+        int attempts = 50;
+        final CountDownLatch inProgressLatch = new CountDownLatch(attempts);
+
+        ActionListener<CreateIndexResponse> indexCreationListener = ActionListener.wrap(response -> {
+            assertTrue(response.isAcknowledged());
+            inProgressLatch.countDown();
+        }, exception -> {
+            if (!(ExceptionsHelper.unwrapCause(exception) instanceof ResourceAlreadyExistsException)) {
+                fail("Failed for reason other than ResourceAlreadyExistsException: " + exception);
+            }
+            inProgressLatch.countDown();
+        });
+
+        ModelIndex modelIndex = ModelIndex.getInstance();
+
+        for (int i = 0; i < attempts; i++) {
+            modelIndex.createModelIndex(indexCreationListener);
+        }
+
+        assertTrue(inProgressLatch.await(30, TimeUnit.SECONDS));
+    }
+
+    public void testModelIndexExists() throws IOException, InterruptedException {
+        ModelIndex modelIndex = ModelIndex.getInstance();
+
+        assertFalse(modelIndex.modelIndexExists());
+
+        final CountDownLatch inProgressLatch = new CountDownLatch(1);
+
+        ActionListener<CreateIndexResponse> indexCreationListener = ActionListener.wrap(response -> {
+            assertTrue(response.isAcknowledged());
+            inProgressLatch.countDown();
+        }, exception -> fail("Model index unable to create"));
+
+        modelIndex.createModelIndex(indexCreationListener);
```

Comment:
is it possible to create modelIndex manually ? if there is bug in createModelIndex, both create and exists will fail because we are using implementation as part of test set up process.

Reply:
Youre right. It is possible. I will add test functions.